### DISCUSSION TITLE
Getting trace id out of a langchain runnable

### DISCUSSION THREAD

**Discussion by salmanmashayekh** [langfuse/langfuse#2342] [Support] [answered]
I have a langchain _runnable_ and I want to instrument it with tracing by creating another _runnable_ that returns the output from the original runnable, as well as the trace id for the langfuse trace. 

Here is what I have:

```
from langfuse import Langfuse
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import AzureChatOpenAI
from langchain_core.runnables import RunnableLambda, RunnablePassthrough, RunnableParallel

langfuse = Langfuse()
prompt = ChatPromptTemplate.from_messages(
    [("human", "Tell me a short joke about {topic}!")]
)
llm = AzureChatOpenAI()

def get_handler():
    trace = langfuse.trace(name="my-trace")
    return trace.get_langchain_handler()

chain = prompt | llm

instrumented_chain = RunnableParallel(
    handler=RunnableLambda(lambda _: get_handler()),
    input=RunnablePassthrough(),
) | RunnableParallel(
    trace_id=RunnableLambda(lambda inputs: inputs["handler"].get_trace_id()),
    chain_output=RunnableLambda(
        lambda inputs: chain.invoke(
            inputs["input"], config={"callbacks": [inputs["handler"]]}
        ).content 
    ),
)

print(instrumented_chain.invoke({"topic": "bear"}))
```

It does work and I get the expected output. However, the trace that I see has these issues:
1. The "Trace" itself (highest level object in the nested object in the UX) has empty input and output (see pic).
2. The tracing related part of the runnable does not appear in the UX.

What am I doing wrong or what's the better/right way to achieve this?

My use case is that I want to know the trace, so that I can add evaluations (through the SDK) later.

![Screenshot 2024-06-13 at 6 14 33 PM](https://github.com/langfuse/langfuse/assets/48693449/ff23f070-7425-462d-b315-d7aaaab458c5)

**Comment by salmanmashayekh** [comment]
Any help is appreciated @marcklingen.

**Reply by GMartin-dev** [reply, marked as answer]
Hi! Did you tried creating the Run id manually before the first runnable? Like this example:
https://github.com/orgs/langfuse/discussions/2172#discussioncomment-9769762

**Reply by salmanmashayekh** [reply]
Thanks for your response @GMartin-dev, but I am not sure how that helps. Are you suggesting to use the langchain run-id as the identifier for the trace (instead of the trace-id)?

**Reply by salmanmashayekh** [reply]
I tried it and that does work! Thanks you @GMartin-dev!

So, if I set the `run_id` when I call the `invoke` method, langfuse uses the same `run_id` for as the `trace_id`.
